### ISSUE TITLE
Linux .deb icon requires packing/making for MacOS

### ISSUE THREAD

**Issue by troywweber7** [electron-userland/electron-builder#6905]
* **Electron-Builder Version**: 22.11.11

* **Node Version**: v16.15.0

* **Electron Version**: v19.0.1
* **Electron Type (current, beta, nightly)**:current

* **Target**: Linux .deb

My `electron-builder.config.json` looks something like this:

```json
{
  "appId": "com.yourdoamnin.yourapp",
  "directories": {
    "buildResources": "resources"
  },
  "files": [
    "assets/**/*",
    "build/**/*",
    "capacitor.config.*",
    "app/**/*"
  ],
  "linux": {
    "target": ["deb"],
    "category": "Education",
    "icon": "dist/.icon-icns/icon.icns"
  },
  "mac": {
    "target": ["zip"],
    "icon": "assets/appIcon.png"
  },
  "win": {
    "icon": "assets/appIcon.ico"
  }
}
```

My `package.json` scripts look like this:

```json
  "scripts": {
    "build": "tsc",
    "electron.start-live": "node ./live-runner.js",
    "electron.start": "npm run build && electron --inspect=5858 ./",
    "electron.pack": "npm run build && electron-builder build -lmw --dir -c ./electron-builder.config.json",
    "electron.make": "npm run build && electron-builder build -l -c ./electron-builder.config.json",
    "postinstall": "electron-builder install-app-deps"
  },
```

`assets/appIcon.png` is a valid file.

If I run `electron.make`, it never generates the .icns file whether I set `linux.icon` to `undefined`, `"assets/appIcon.png"`, or `"dist/.icon-icns/icon.icns"`. So when I install the generated .deb file, it is either the blank gear icon or the default electron icon.

However, if I run `electon.pack`, it will create the .icns file because it builds for mac, then if I set `linux.icon` to `"dist/.icon-icns/icon.icns"` and run `electron.make` again, the installed .deb file will have the correct icon. Why should I have to pack it for mac in order for the linux debian icon to work? This seems like a bad workflow.

**Comment by burgil** [comment]
Well, I think the source of this issue is just confusion, I could be wrong but as far as I'm aware, and I did it before:

1) You can not use a `.png` as mac icons:
![image](https://user-images.githubusercontent.com/41600149/172071598-b81e8239-d376-404f-b9ab-13b9648b5977.png)

2) (And don't catch me on this) Electron builder does not generate the icons for you, You have to pre generate them (only once)  [just once] - and put them wherever needed, Electron Builder simply use the files you specify in that config, do change it to something like this:

```
{
  "appId": "com.yourdoamnin.yourapp",
  "directories": {
    "buildResources": "resources"
  },
  "files": [
    "assets/**/*",
    "build/**/*",
    "capacitor.config.*",
    "app/**/*"
  ],
  "linux": {
    "target": ["deb"],
    "category": "Education",
    "icon": "dist/.icon-icns"
  },
  "mac": {
    "target": ["zip"],
    "icon": "assets/appIcon.icns"
  },
  "win": {
    "icon": "assets/appIcon.ico"
  }
}
```
and make sure all of the above are valid path it looks weird that mac and win is using assets and linux is using dist in my opinion, but again I am no expert just another guy like you

**Comment by burgil** [comment]
> You are saying to point mac.icon to ./assets/appIcon.png one time, run electron-builder to let it create the ./dist/.icon-icns/icon.icns at least once, move that icon.icns to ./assets/appIcon.icns, then finally change mac.icon and linux.icon to point to the new ./assets/appIcon.icns?

Yes that is what I meant I guess, I just used an online icon generator lol

> Any thoughts or insights from someone with more experience (than @burgil and myself)

Haha I agree, I'm not a contributor just trying to help :)

> Building for MacOS will automatically generate the icon.icns

I did not know that, thank you I've learned something new 👍 

> Also, to your point that You cannot use a .png as mac icons, [this documentation seems to disagree with you](https://www.electron.build/icons#macos).

Well I see that just now, I guess that I didn't know they were generating that icons in that case,

I would like to point out that when I made an app for Mac once using xCode it did not let me put a png as the favicon, and I had to generate the .icns manually using a terminal command that was only available in mac os to convert a png to icns, did that years ago

> So why, when building only for Linux, shouldn't the default be to generate a .icns just as MacOS would, then generate the required Linux icons from that?

Are you building from Windows, Mac or Linux?

this article talks about cross platform builds and how you can only build from mac to windows but not vice versa I think..

![image](https://user-images.githubusercontent.com/41600149/172233472-949b2f81-f7d1-42e9-9f76-108764e9a1ea.png)

https://www.electron.build/multi-platform-build

> Free public [Electron Build Service](https://github.com/electron-userland/electron-build-service) is used to build Electron app for Linux on Windows. On macOS/Linux you can build Electron app for Windows locally, except Appx for Windows Store (in the future (feel free to file issue) electron-build-service will support Appx target).

## Anywho good luck 😄

**Comment by burgil** [comment]
Well I researched a lot about that topic actually and I believe the answer is that you can only generate a mac build from mac, or through a build server (that is running mac lol), So I believe the answer is that you can't generate a mac build from linux

**Comment by troywweber7** [comment]
@burgil yep, that's my conclusion as well.

But just to be super clear for anyone who comes across this issue, while this extra information about building for MacOS is helpful, it is not relevant to my original question.

On Linux you can build the unpacked MacOS files with `--dir`, which generates the `.icns`, which the build for Linux can then use to generate its PNG files. But if you want to generate only for Linux, there is no option to leave out the MacOS dir build and generate all the necessary Linux icons automatically from a single PNG.

There are workarounds, but obviously they're not an ideal workflow.

**Comment by benbucksch** [comment]
Also, please don't close issues just because they have no comments. Bugs are still bugs, whether people actively whine about it not. It seems this auto-closing is a new thing, given that this bug is open since 2022 - and unfortunately still not fixed.
